code/modules/client/preference_setup/general/04_equipment.dm

```diff
@@ -48,70 +48,98 @@
 
 /datum/category_item/player_setup_item/general/equipment/gather_save_parameters()
 	return list(
-		"underwear" = pref.underwear,
-		"undershirt" = pref.undershirt,
-		"socks" = pref.socks,
+		"all_underwear" = json_encode(pref.all_underwear),
+		"all_underwear_metadata" = json_encode(pref.all_underwear_metadata),
 		"backbag" = pref.backbag,
 		"backbag_style" = pref.backbag_style,
 		"id" = pref.current_character,
 		"ckey" = PREF_CLIENT_CKEY
 	)
 
 /datum/category_item/player_setup_item/general/equipment/sanitize_character(var/sql_load = 0)
+	if(!istype(pref.all_underwear))
+		pref.all_underwear = list()
+
+		for(var/datum/category_group/underwear/WRC in global_underwear.categories)
+			for(var/datum/category_item/underwear/WRI in WRC.items)
+				if(WRI.is_default(pref.gender ? pref.gender : MALE))
+					pref.all_underwear[WRC.name] = WRI.name
+					break
+
+	if(!istype(pref.all_underwear_metadata))
+		pref.all_underwear_metadata = list()
+
+	for(var/underwear_category in pref.all_underwear)
+		var/datum/category_group/underwear/UWC = global_underwear.categories_by_name[underwear_category]
+		if(!UWC)
+			pref.all_underwear -= underwear_category
+		else
+			var/datum/category_item/underwear/UWI = UWC.items_by_name[pref.all_underwear[underwear_category]]
+			if(!UWI)
+				pref.all_underwear -= underwear_category
+
+	for(var/underwear_metadata in pref.all_underwear_metadata)
+		if(!(underwear_metadata in pref.all_underwear))
+			pref.all_underwear_metadata -= underwear_metadata
+
 	if (sql_load)
 		pref.backbag = text2num(pref.backbag)
+		pref.backbag_style = text2num(pref.backbag_style)
+		if(istext(pref.all_underwear))
+			var/before = pref.all_underwear
+			try
+				pref.all_underwear = json_decode(pref.all_underwear)
+			catch(var/exception/e)
+				log_debug("UNDERWEAR: Caught [e]. Initial value: [before]")
+				pref.all_underwear = list()
+		if(istext(pref.all_underwear_metadata))
+			var/before = pref.all_underwear_metadata
+			try
+				pref.all_underwear_metadata = json_decode(pref.all_underwear_metadata)
+			catch(var/exception/e)
+				log_debug("UNDERWEAR METADATA: Caught [e]. Initial value: [before]")
+				pref.all_underwear_metadata = list()

```

Comment:
This is way too late in the chain. It should be the first thing done in this proc, because all of the sanitization requires that these be deserialized already.